packages/flutter/lib/src/material/data_table.dart

```diff
@@ -847,6 +852,8 @@ class DataTable extends StatelessWidget {
     final bool anyRowSelectable = rows.any((DataRow row) => row.onSelectChanged != null);
     final bool displayCheckboxColumn = showCheckboxColumn && anyRowSelectable;
     final bool allChecked = displayCheckboxColumn && !rows.any((DataRow row) => row.onSelectChanged != null && !row.selected);
```

Comment:
If a table has tens of thousands of rows then scanning all of the rows 2X (worst case) seems like a bad idea.

Reply:
I optimized this now to:
```dart
final Iterable<DataRow> rowsChecked = displayCheckboxColumn ?
  rows.where((DataRow row) => row.onSelectChanged != null && row.selected) : <DataRow>[];
final bool allChecked = displayCheckboxColumn && rowsChecked.length == rows.length;
final bool anyChecked = displayCheckboxColumn && rowsChecked.isNotEmpty;
```